src/SmoFacade/Server.cs

```diff
@@ -136,61 +136,51 @@ public Database Database(string dbName)
     ///   TODO: we should support a backup path somehow with configuration
     /// </summary>
     /// <param name="databaseName"></param>
-    /// <param name="backupPathTemplate">
+    /// <param name="backupDirectoryPath">
     ///   A template string for backup location:
     ///   {0} databaseName
     ///   {1} time
     /// </param>
-    public void LogBackup(string databaseName, string backupPathTemplate = null)
+    public void LogBackup(string databaseName, string backupDirectoryPath = null)
     {
-      backupPathTemplate = backupPathTemplate ?? DefaultBackupPathTemplate(BackupFileTools.BackupType.Log);
       var backup = new Backup
         { Action = BackupActionType.Log, Database = databaseName, LogTruncation = BackupTruncateLogType.Truncate };
-      Backup(backup, backupPathTemplate);
+      Backup(backup, backupDirectoryPath, databaseName, BackupFileTools.BackupType.Log);
     }
 
     /// <summary>
     ///   Generate a full backup, not truncating the transaction log, and storing it in the default backup destination.
     ///   TODO: we should support a backup path somehow with configuration
     /// </summary>
     /// <param name="databaseName">name of the database to backup</param>
-    /// <param name="backupPathTemplate">
+    /// <param name="backupDirectoryPath">
     ///   A template string for backup location:
     ///   {0} databaseName
     ///   {1} time
     /// </param>
-    public void FullBackup(string databaseName, string backupPathTemplate = null)
+    public void FullBackup(string databaseName, string backupDirectoryPath = null)
     {
-      backupPathTemplate = backupPathTemplate ?? DefaultBackupPathTemplate(BackupFileTools.BackupType.Full);
       var backup = new Backup {
         Action = BackupActionType.Database, Database = databaseName, LogTruncation = BackupTruncateLogType.NoTruncate
       };
-      Backup(backup, backupPathTemplate);
+      Backup(backup, backupDirectoryPath, databaseName, BackupFileTools.BackupType.Full);
     }
 
-    private void Backup(Backup backup, string backupPathTemplate)
+    private void Backup(Backup backup, string backupDirectoryPath, string databaseName, BackupFileTools.BackupType type)
     {
-      var filePath = string.Format(backupPathTemplate,
-                                   backup.Database,
-                                   DateTime.Now.ToString("yyyy_MM_dd_hhmmss_fff"));
+      backupDirectoryPath = backupDirectoryPath ?? _server.BackupDirectory;
+      if(backupDirectoryPath.EndsWith("\\") || backupDirectoryPath.EndsWith("/"))
+        backupDirectoryPath = backupDirectoryPath.Substring(0, backupDirectoryPath.Length - 1);
+
+      var filePath =
+        $"{backupDirectoryPath}/{databaseName}/{databaseName}_backup_{DateTime.Now.ToString("yyyy_MM_dd_hhmmss_fff")}.{BackupFileTools.BackupTypeToExtension(type)}";

```

Comment:
The template allowed for the databaseName in the directory path as well as the file name, but this new implementation forces the directory per database setup and our naming conventions on all consumers.

Reply:
From an interface perspective, there is no way to know what kind of template is required.  The only way to know how many template fields and what they are used for is to look at the source code.
I can remove the `"database name as a directory"`, but I'd really prefer to not have a templated string interface

Reply:
Ya, the templated string interface is pretty poor. It's just so easy.

Now that we're moving the directory path lookup from C# to SQL, should we just have the query generate the full backup and file path? I feel like even making the file name be databasename_backup_timestamp.bak is pretty prescriptive. If we're running within the context the database we're managing everything needed should be there.

Reply:
I didn't do that because `_server.BackupDirectory` isn't going to return a templated date/filename.... although it would be nice to have control over the filename as well... hmm